Library/Homebrew/command.rb

```diff
@@ -0,0 +1,321 @@
+        @valid_options.select { |h| /^-(?!-)/ =~ h[:option] || h[:switch] }
+                      .map { |h| h[:switch] || h[:option] }
+      end
+
+      def option(*args, **option_hash, &block)
+        if args.length == 1
+          if args[0].length == 1
+            option_hash[:switch] = args[0]
+          else
+            option_hash[:option] = args[0]
+          end
+        end
+
+        option_hash[:option] = "--#{option_hash[:option]}" if option_hash[:option]
+        option_hash[:switch] = "-#{option_hash[:switch]}" if option_hash[:switch]
+        if option_hash[:option].nil?
+          option_hash[:option] = option_hash[:switch]
+          option_hash[:switch] = nil
+        end
+        option_name = option_hash[:option]
+        if @parent.nil?
+          @root_options.push(option_name)
+        else
+          hash = @valid_options.find { |x| x[:option] == @parent }
+          if hash[:child_options].nil?
+            hash[:child_options] = [option_name]
+          else
+            hash[:child_options].push(option_name)
+          end
+        end
+        add_valid_option(option_hash)
+        return unless block_given?
+        old_parent = @parent
+        @parent = option_name
+        instance_eval(&block)
+        @parent = old_parent
+      end
+
+      def desc(desc)
+        if @description.nil?
+          @description = desc
+        else
+          @description = <<-EOS.undent
+            #{@description}
+            #{desc}
+          EOS
+                               .strip
+        end
+      end
+
+      def argv_invalid_options_passed(argv)
+        argv_options_only = argv.select { |arg| /^--/ =~ arg }
+        argv_options_only = argv_options_only.uniq
+        valid_option_names =
+          @valid_options
+          .map { |option_hash| option_hash[:option] }
+
+        argv_options_only
+          .reject { |opt| valid_option_names.include?(opt.split("=", 2)[0]) }
+          .map { |opt| opt.split("=", 2)[0] }
+      end
+
+      def argv_options_without_value_passed(argv)
+        argv_options_only = argv.select { |arg| /^--/ =~ arg }
+        valid_options_with_values =
+          @valid_options
+          .select { |option_hash| option_hash[:value] }
+
+        options_without_value =
+          argv_options_only
+          .select { |opt| valid_options_with_values.map { |x| x[:option] }.include?(opt.split("=", 2)[0]) }
+          .select do |opt|
+            (opt.include?("=") && opt.split("=", 2)[1].empty?) ||
+              (!opt.include?("=") && (argv[argv.index(opt)+1].nil? ||
+                            argv[argv.index(opt)+1].start_with?("-")))
+          end
+        options_without_value.map do |opt|
+          opt_name = opt.split("=", 2)[0]
+          opt_value = @valid_options.find { |x| x[:option] == opt_name }[:value]
+          [opt_name, opt_value]
+        end
+      end
+
+      def argv_invalid_switches_passed(argv)
+        argv.select { |arg| /^-(?!-)/ =~ arg }
```

Comment:
@MikeMcQuaid @woodruffw, `brew style` shows a warning here: `Use reject instead of inverting select`. Thats a false warning, right?